web-ui/src/main/resources/catalog/components/viewer/projectionSwitcher/ProjectionSwitcher.js

```diff
@@ -0,0 +1,175 @@
+ * your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
+
+(function() {
+  goog.provide('gn_projectionswitcher');
+
+  var module = angular.module('gn_projectionswitcher', [
+  ]);
+
+
+  /**
+   * @ngdoc directive
+   * @name gn_projectionswitcher.directive:gnProjectionSwitcher
+   * 
+   * @description The `gnProjectionSwitcher` directive adds a button to the map
+   *              that allows you to change the projection.
+   */
+  module.directive('gnProjectionSwitcher', [
+    'gnViewerSettings',
+    function(gnViewerSettings) {
+      return {
+        restrict: 'A',
+        templateUrl: '../../catalog/components/viewer/projectionSwitcher/' +
+            'partials/projectionSwitcher.html',
+        scope: {
+          map: '=gnProjectionSwitcher'
+        },
+        controllerAs: 'gnProjectionSwitcherCtrl',
+        controller: ['$scope', 'gnViewerSettings', 'gnMap', '$rootScope', 
+          function(scope, gnViewerSettings, gnMap, $rootScope) {
+
+          scope.changeLayerProjection = function(
+            layer, oldProj, newProj) {
+            if (layer instanceof ol.layer.Group) {
+              layer.getLayers().forEach(
+                  function(subLayer) {
+                    this.changeLayerProjection(
+                      subLayer, oldProj,
+                      newProj);
+                  });
+            } else if (layer instanceof ol.layer.Tile) {
+              var tileLoadFunc = layer.getSource()
+                .getTileLoadFunction();
+              layer.getSource().setTileLoadFunction(
+                tileLoadFunc);
+            } else if (layer instanceof ol.layer.Vector) {
+              var features = layer.getSource().getFeatures();
+              for (var i = 0; i < features.length; i += 1) {
+                features[i].getGeometry().transform(
+                  oldProj, newProj);
+              }
+            }
+          };
+          
+          scope.switchProjection = function(projection) {
+              var view = scope.map.getView();
+              var oldProj = view.getProjection();
+              var newProj = ol.proj.get(projection);
+              
+              if(oldProj == newProj) {
+                return;
+              }
+              
+              var projectionConfig = {};
+              
+              $.each(gnViewerSettings.mapConfig.switcherProjectionList, function(i, config) {
```

Comment:
We should avoid using jQuery where unnecessary; here you can either do a for loop or use `forEach` on the array.

Reply:
We have been using that $.each in GN for years... do we have guidelines about this? I always try to copy the development style of the code around it, but not sure if we have agreed on something global.

Reply:
Good question, I was referring to general practices when coding in JS. jQuery has been heavily used for years until people realized that most of its functions were added in the native language eventually.

We definitely need coding guidelines in JS and Java somewhere...

Reply:
Yes please, I feel sometimes lost trying to decide what convention to use so I end up trying to copy code around... which may be old.